### ISSUE TITLE
Re-authenticating using new AD B2C Server flow (while using local debugging)

### ISSUE THREAD

**Issue by ltaljaard** [Azure/azure-mobile-apps-net-client#486]
The TodoAzureAuthADB2CServerFlow example: [https://developer.xamarin.com/samples/xamarin-forms/WebServices/TodoAzureAuthADB2CServerFlow/](url)

is the most up to date one that I could find. But it does not include steps on how to handle automatic re-authentication. I'm trying to get this working while using local debugging, something along these lines:

[https://social.msdn.microsoft.com/Forums/en-US/ef8b58a7-922e-4190-b19b-45d35cc8aa5b/local-debug-aad-authentication?forum=azuremobile](url)

Or [http://www.systemsabuse.com/2015/12/04/local-debugging-with-user-authentication-of-an-azure-mobile-app-service/](url)

The above links are a bit dated by now but that is all I can find on the net. And this used to work fine for me in the past.

Essentially I think it comes down to how to handle an "unauthorized" (401) error inside one's own overridden SendAsync method in your own DelegatingHandler-derived class (this is what is not included in any recent examples or tutorials). It might also have to do with the AlternateLoginHost (part of this azure-mobile-apps-net-client repo), I am not sure (no docs!).

I am logging this as a bug because it (auto re-authentication) used to work fine for me but the recent changes to the AD B2C flows have broken something, or some or other parameter or setting must have changed. This is such a complicated scenario that I think it would be best explained by a working example. But that said I also need urgent help to get it to work so can't wait for the official sample/s to be eventually updated.

**Comment by jwallra** [comment]
Hi @ltaljaard - The most up-to-date documentation we have concerning debugging authentication flows can be found here: https://adrianhall.gitlab.io/develop-mobile-apps-with-csharp-and-azure/chapter2/debugging/

The rest of our documentation is [here](https://docs.microsoft.com/en-us/azure/app-service-mobile/).

If those resources don't help, could you list the SDKs you're using, their versions, and repro steps?  If you can, it would be most helpful to have a sample project that reproduces the issue, although I think based on what you've said so far that it's likely there's some environment issue.

**Comment by MatkovIvan** [comment]
@ltaljaard UI thread issue can be simply resolved by [`InvokeOnMainThread`](https://docs.microsoft.com/en-us/xamarin/ios/user-interface/ios-ui/ui-thread)

**Comment by MatkovIvan** [comment]
@ltaljaard Did you manage to make it work? Are there any other problems besides `UIKitThreadAccessException` in that example?

**Comment by ltaljaard** [comment]
Hi Ivan, I have started to try it out and it’s at least getting onto the Azure B2C logon page now. This is a potential fix on iOS, I’ll have to look at the Android case as well.

I wonder why it has only started requiring the execution on the UI thread now whereas it worked fine in the past? The docs should make mention of this and incorporate it into the samples.

Will let you know the final outcome, please keep issue open for now.

On 11 Jan 2019, at 09:52, Ivan Matkov <notifications@github.com> wrote:

@ltaljaard <https://github.com/ltaljaard> Did you manage to make it work? Are there any other problems besides UIKitThreadAccessException in that example?

**Comment by ltaljaard** [comment]
@MatkovIvan Ok so within method LoginAsync of class iOSLoginProvider of Adrian Hall's Chapter 2 example, which can be found here: [https://gitlab.com/adrianhall/develop-mobile-apps-with-csharp-and-azure/blob/master/Chapter2/TaskList/TaskList.iOS/Services/iOSLoginProvider.cs](https://gitlab.com/adrianhall/develop-mobile-apps-with-csharp-and-azure/blob/master/Chapter2/TaskList/TaskList.iOS/Services/iOSLoginProvider.cs)

I have changed it from this:

`return await client.LoginAsync(RootView, "aad");`

to this:

`MobileServiceUser user = null;`
`UIApplication.SharedApplication.InvokeOnMainThread(async () => {
    user = await client.LoginAsync(RootView, "aad");
});`
`return user;`

What happens now is that it is not awaiting the call to LoginAsync, it just continues on and ends up returning null in the user variable. But once the iOSLoginProvider.LoginAsync() method completes, the MobileServiceClient.LoginAsync() does actually get executed because it then navigates to the Azure B2C logon page. but by then it is too late, I need to be able to capture the "user" return value. So the above code is only starting the MobileServiceClient.LoginAsync() task whereas it should be awaiting it.

Could you advise how to code it so it awaits properly? I have tried several things but without success.

**Comment by ltaljaard** [comment]
@MatkovIvan Ok great, another step closer. That code now brings up the Azure B2C login page (login.microsoftonline.com) in a browser and when the user supplies a valid user & pw and clicks on Sign In then the control returns to the code as expected. The new token is then available in variable "user". So the call to  MobileServiceClient.LoginAsync() is now properly awaited in the case of a successful login.

But when the user clicks the Done button (top left) on the login.microsoftonline.com browser page in order to cancel the login, i.e. instead of clicking on Sign In, then the control does not return to my code. I have a catch(Exception) block over everything but it does not catch an exception for the case there the user cancels the login by clicking on the Done button.

**Comment by MatkovIvan** [comment]
Ok, small addition:

```C#
var user = new TaskCompletionSource<MobileServiceUser>();
UIApplication.SharedApplication.InvokeOnMainThread(async () =>
{
    try
    {
        user.SetResult(await client.LoginAsync(RootView, "aad"));
    }
    catch (Exception e)
    {
        user.SetException(e);
    }
});
return await user.Task;
```

**Comment by ltaljaard** [comment]
@MatkovIvan Thanks, just tried it, but no luck. Nothing gets caught if user clicks the done button. It just goes back to the main app at that point. I can show you over teamviewer if that would help.

**Comment by MatkovIvan** [comment]
I've reproduced it, it seems this is an SDK issue.
The problem is that the completion block isn't executed here:
https://github.com/Azure/azure-mobile-apps-net-client/blob/6ecc4185b3a7c366881dc805ae4b6187a26850df/src/Microsoft.Azure.Mobile.Client/Platforms/ios/Authentication/MobileServiceUIAuthentication.cs#L94

I'm not sure when it broke, probably this behavior changed in one of the latest iOS.

**Comment by ltaljaard** [comment]
@MatkovIvan Thank you for fixing the underlying issue. A reminder to include this (re-authentication upon expired token) in the regression tests.

When can we expect to get this in the stable Nuget release? And when in the next pre-release?

**Comment by MatkovIvan** [comment]
@ltaljaard Sorry for the delay. The fix is available in `4.1.2-prerelease.481290` beta nuget.
